pkg/pool-hooks/contracts/StableSurgeHookExample.sol

```diff
@@ -0,0 +1,235 @@
+    PoolSwapParams,
+    HookFlags,
+    SwapKind
+} from "@balancer-labs/v3-interfaces/contracts/vault/VaultTypes.sol";
+import { VaultGuard } from "@balancer-labs/v3-vault/contracts/VaultGuard.sol";
+import { BaseHooks } from "@balancer-labs/v3-vault/contracts/BaseHooks.sol";
+
+import { StableMath } from "@balancer-labs/v3-solidity-utils/contracts/math/StableMath.sol";
+import { FixedPoint } from "@balancer-labs/v3-solidity-utils/contracts/math/FixedPoint.sol";
+
+/**
+ * @notice Hook that applies a fee for out of range or undesirable amounts of tokens in relation to a threshold.
+ * @dev Uses the dynamic fee mechanism to apply a directional fee.
+ */
+contract StableSurgeHookExample is BaseHooks, VaultGuard, Ownable {
+    using FixedPoint for uint256;
+    // Only pools from a specific factory are able to register and use this hook.
+    address private immutable _allowedFactory;
+    // Defines the range in which surging will not occur; meaning surging will occur when:
+    // B(i-after swap) / (sum(B(n-after swap))) > 1/n + gamma
+    uint256 private _threshold;
+    // An amplification coefficient to amplify the degree a fee increases upon the threshold being triggered
+    uint256 private _surgeCoefficient;
+
+    /**
+     * @notice A new `StableSurgeHookExample` contract has been registered successfully.
+     * @dev If the registration fails the call will revert, so there will be no event.
+     * @param hooksContract This contract
+     * @param factory The factory (must be the allowed factory, or the call will revert)
+     * @param pool The pool on which the hook was registered
+     */
+    event StableSurgeHookExampleRegistered(
+        address indexed hooksContract,
+        address indexed factory,
+        address indexed pool
+    );
+
+    /**
+     * @notice The threshold has been changed in a `StableSurgeHookExample` contract.
+     * @dev Note, the initial threshold is set on deployment and an event is emitted.
+     * @param hooksContract This contract
+     * @param threshold The new threshold
+     */
+    event ThresholdChanged(address indexed hooksContract, uint256 indexed threshold);
+
+    /**
+     * @notice The surgeCoefficient has been changed in a `StableSurgeHookExample` contract.
+     * @dev Note, the initial surgeCoefficient is set on deployment and an event is emitted.
+     * @param hooksContract This contract
+     * @param surgeCoefficient The new surgeCoefficient
+     */
+    event SurgeCoefficientChanged(address indexed hooksContract, uint256 indexed surgeCoefficient);
+
+    constructor(
+        IVault vault,
+        address allowedFactory,
+        uint256 threshold,
+        uint256 surgeCoefficient
+    ) VaultGuard(vault) Ownable(msg.sender) {
+        _allowedFactory = allowedFactory;
+        _threshold = threshold;
+        _surgeCoefficient = surgeCoefficient;
+        emit ThresholdChanged(address(this), threshold);
+        emit SurgeCoefficientChanged(address(this), surgeCoefficient);
```

Comment:
What we typically do here is encapsulate the event with the set function. 
```suggestion
        _setThreshold(threshold);
        _setSurgeCoefficient(surgeCoefficient);
```
Then, for both the threshold and the surgeCoefficient:

```
function setThreshold(uint64 newThreshold) external onlyOwner {
    _setThreshold(newThreshold);
}

function _setThreshold(uint64 newThreshold) private {
    // This should be validated; e.g. <= FixedPoint.ONE or some max value?
    _threshold = newThreshold;

    emit ThresholdChanged(address(this), newThreshold);
}
```

Reply:
How is surge coefficient validated? (i.e., what is the range)? You can't go over 100% on a dynamic fee, or the Vault will revert.

Reply:
Will check with Zen on best way to validate.

Reply:
From Zen:
* I don’t think the threshold should matter so much as long as it’s less than 1 - 1/n 
* surge coefficient: baseFee * 𝜇 / (1/n) < 100 would be the test then